src/data/pi/lex.lean

```diff
@@ -44,12 +44,28 @@ basic API, just in case -/
 notation `Πₗ` binders `, ` r:(scoped p, lex (Π i, p i)) := r
 
 @[simp] lemma to_lex_apply (x : Π i, β i) (i : ι) : to_lex x i = x i := rfl
-@[simp] lemma of_lex_apply (x : lex (Π i, β i)) (i : ι) : of_lex x i = x i := rfl
+@[simp] lemma of_lex_apply (x : Πₗ i, β i) (i : ι) : of_lex x i = x i := rfl
 
-instance [has_lt ι] [Π a, has_lt (β a)] : has_lt (lex (Π i, β i)) := ⟨pi.lex (<) (λ _, (<))⟩
+lemma is_trichotomous_lex [∀ i, is_trichotomous (β i) s] (wf : well_founded r) :
+  is_trichotomous (Π i, β i) (pi.lex r @s) :=
+{ trichotomous := λ a b,
+    begin
+      by_cases h : ∃ i, a i ≠ b i,
+      { let i := wf.min _ h,
+        have hri : ∀ j, r j i → a j = b j,
+        { intro j, rw ← not_imp_not,
+          exact λ h', wf.not_lt_min _ _ h' },
+        have hne : a i ≠ b i, from wf.min_mem _ h,
+        cases (trichotomous (a i) (b i) : s (a i) (b i) ∨ a i = b i ∨ s (b i) (a i)) with hi hi,
+        exacts [or.inl ⟨i, hri, hi⟩,
+          or.inr $ or.inr $ ⟨i, λ j hj, (hri j hj).symm, hi.resolve_left hne⟩] },
+      { push_neg at h, exact or.inr (or.inl (funext h)) }
```

Comment:
Isn't it (a bit) faster to case on `a = b` and use `function.ne_iff`?

Reply:
Done, thanks!